vllm/model_executor/models/ultravox.py

```diff
@@ -140,78 +92,99 @@ def input_mapper_for_ultravox(ctx: InputContext, data: object):
+
+        def get_replacement_ultravox(item_idx: int):
+            audio_data, sr = mm_items.audio[item_idx]
+            audio_data = self._resample_audio(audio_data, sr)["audio"]
+            audio_len = audio_data.shape[0]
+            nb_encoder_frames = int(round(audio_len / encoder_ds_factor +
+                                          1e-4))
+            audio_token_len = int(np.ceil(nb_encoder_frames / stack_factor))
+            max_audio_token_len = get_ultravox_max_audio_tokens(self.ctx)
+            audio_token_len = min(max(1, audio_token_len), max_audio_token_len)
+            return [_AUDIO_PLACEHOLDER_TOKEN] * audio_token_len
+
+        return [
+            PromptReplacement(
+                modality="audio",
+                target="<|audio|>",
+                replacement=get_replacement_ultravox,
+            )
+        ]
 
-    feature_extractor = whisper_feature_extractor(ctx)
-    audios = multi_modal_data["audio"]
-    if not isinstance(audios, list):
-        audios = [audios]
-
-    audio_token_counts = []
-    for audio in audios:
-        if isinstance(audio, torch.Tensor):
-            audio_num_tokens = audio.shape[1]
-            audio_token_counts.append(audio_num_tokens)
-        else:
-            audio_data, sample_rate = audio
-            audio_length = audio_data.shape[0]
-            if sample_rate != feature_extractor.sampling_rate:
-                # Account for resampling.
-                adjustment = feature_extractor.sampling_rate / sample_rate
-                audio_length = math.ceil(adjustment * audio_length)
-
-            feature_extractor_output_length = math.ceil(
-                (audio_length - (feature_extractor.hop_length - 1)) /
-                feature_extractor.hop_length)
-
-            uv_config = ctx.get_hf_config(UltravoxConfig)
-            audio_num_tokens = min(
-                max(
-                    1,
-                    math.ceil(feature_extractor_output_length /
-                              (uv_config.stack_factor * 2))),
-                get_ultravox_max_audio_tokens(ctx))
-            audio_token_counts.append(audio_num_tokens)
-
-    tokenizer = cached_get_tokenizer(ctx.model_config.tokenizer)
-
-    new_prompt, new_token_ids, ranges = repeat_and_pad_placeholder_tokens(
-        tokenizer,
-        inputs.get("prompt"),
-        inputs["prompt_token_ids"],
-        placeholder_token_id=_AUDIO_PLACEHOLDER_TOKEN,
-        repeat_count=audio_token_counts,
-    )
-
-    # NOTE: Create a defensive copy of the original inputs
-    return token_inputs(prompt_token_ids=new_token_ids,
-                        prompt=new_prompt,
-                        multi_modal_data=multi_modal_data,
-                        multi_modal_placeholders={"audio": ranges})
+    def _get_dummy_mm_inputs(
+        self,
+        mm_counts: Mapping[str, int],
+    ) -> ProcessorInputs:
+        feature_extractor = whisper_feature_extractor(self.ctx)
```

Comment:
```suggestion
        feature_extractor = self._get_hf_processor().feature_extractor
```

This should work since it's part of the HF processor (https://github.com/huggingface/transformers/blob/main/src/transformers/models/whisper/processing_whisper.py)